### PR TITLE
vstart.sh: make sure that --localhost does only bind to 127.0.0.1 in all cases 

### PR DESCRIPTION
**Pull request by leonid-s-usov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
  - [x] New feature (ticket optional)
  - [x] No impact that needs to be tracked
  - [x] No doc update is appropriate
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by leonid-s-usov** [comment]
@batrick please review this change.

File: src/vstart.sh [hunk 1]
```diff
@@ -732,6 +734,12 @@ prepare_conf() {
         msgr_conf="ms bind msgr2 = false
                    ms bind msgr1 = true"
     fi
+    if [ $force_addr -eq 1 ]; then
+        msgr_conf+="
+                    public bind addr = $IP
+                    public addr = $IP
+                    cluster addr = $IP"
+    fi
```
Comment (batrick) [inline src/vstart.sh:742]: Try:

```
diff --git a/src/vstart.sh b/src/vstart.sh
index d91c2d70331..1d54f28185e 100755
--- a/src/vstart.sh
+++ b/src/vstart.sh
@@ -742,9 +742,11 @@ prepare_conf() {
     if [ $force_addr -eq 1 ]; then
         msgr_conf+="
                     public bind addr = $IP
-                    public addr = $IP
-                    cluster addr = $IP"
+                    public addr = $IP"
     fi
+    msgr_conf+="
+                cluster addr = 127.0.0.1
+                cluster network = 127.0.0.1/8"
 
     wconf <<EOF
 ; generated by vstart.sh on `date`
```

I think we can hard-code `cluster addr` and `cluster network` to always use loopback.

File: src/vstart.sh [hunk 2]
```diff
@@ -901,6 +909,26 @@ EOF
...
+    #   // check if the public_bind_addr option is set
+    #  if (!g_conf()->public_bind_addr.is_blank_ip()) {
+    #    bind_addrs = make_mon_addrs(g_conf()->public_bind_addr);
+    #  }
+    #
+    if [ $force_addr -eq 1 ]; then
+        wconf <<EOF
+        ; this is to counter the explicit public_bind_addr in the [global] section
+        ; see src/vstart.sh for more info
+        public bind addr =
```
Comment (batrick) [inline src/vstart.sh:929]: Interesting. Maybe a tracker ticket to fix that would be appropriate.

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/vstart.sh (hunk 1 above)
Comment (leonid-s-usov) [reply src/vstart.sh]: done

Comment (batrick) [reply src/vstart.sh]: The API test failure may be related to this, please check.

Comment (leonid-s-usov) [reply src/vstart.sh]: yeah, that's likely. I'll revert this for now.

**Review by batrick — APPROVED** [review]

**Comment by leonid-s-usov** [comment]
@batrick who else could review this to get approved?

**Comment by batrick** [comment]
> @batrick who else could review this to get approved?

I've pinged the dashboard folks on slack. Changes to vstart make break dashboard tests in surprising ways so they like to chime in on these.

**Review by nizamial09 — APPROVED** [review]
dashboard works! i'll trigger the ceph dashboard e2e (which runs on top of mstart) just in case to check whether the mstart cluster starts or not in the CI.

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
[unrelated failure](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/14028/consoleFull#99679487573a8703b-5adb-41c5-84a0-8cf4e065ba3d) in dashboard e2e but it looks like the test suite is working so all good from us!

**Comment by leonid-s-usov** [comment]
Thank you all 🙏🏻

**Pull request merged**